File: cozy-desktop/windows/CozyDesktop/Services/PeerCoordinator.cs
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;

namespace CozyDesktop.Services;

/// <summary>
/// A tiny control channel between Cozy desktops (PC ↔ Mac) for multi-controller
/// handoff. Each instance listens on a control port and also dials the peer. When
/// a machine's user becomes active, it broadcasts CLAIM; peers receiving it yield
/// and become clients of the claimer.
///
/// Line protocol (newline-terminated):  HELLO &lt;name&gt;   |   CLAIM &lt;name&gt;
/// </summary>
public sealed class PeerCoordinator : IDisposable
{
    public const int ControlPort = 24811;

    public event Action<string, string>? PeerClaimedControl; // (peerName, peerIp)
    public event Action<string>? Log;

    private TcpListener? _listener;
    private readonly List<TcpClient> _conns = new();
    private CancellationTokenSource? _cts;
    private string _myName = "";
    private string? _peerIp;

    public void Start(string myName, string? peerIp)
    {
        _myName = myName;
        _peerIp = string.IsNullOrWhiteSpace(peerIp) ? null : peerIp;
        _cts = new CancellationTokenSource();
        _listener = new TcpListener(IPAddress.Any, ControlPort);
        try { _listener.Start(); } catch (Exception ex) { Log?.Invoke($"coordinator listen failed: {ex.Message}"); }
        _ = AcceptLoop(_cts.Token);
        _ = ConnectLoop(_cts.Token);
    }

    private async Task AcceptLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested && _listener != null)
        {
            try { AddConn(await _listener.AcceptTcpClientAsync(ct)); }
            catch { break; }
        }
    }

    // Keep a connection to the configured peer alive (retry every few seconds).
    private async Task ConnectLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            bool connected;
            lock (_conns) connected = _conns.Any(c => c.Connected);
            if (_peerIp != null && !connected)
            {
                try
                {
                    var c = new TcpClient();
                    await c.ConnectAsync(_peerIp, ControlPort, ct);
                    AddConn(c);
                }
                catch { /* peer not up yet */ }
            }
            try { await Task.Delay(3000, ct); } catch { break; }
        }
    }

    private void AddConn(TcpClient c)
    {
        lock (_conns) _conns.Add(c);
        SendLine(c, $"HELLO {_myName}");
        _ = ReadLoop(c);
    }

    private async Task ReadLoop(TcpClient c)
    {
        try
        {
            using var reader = new StreamReader(c.GetStream(), Encoding.UTF8);
            string? line;
            while ((line = await reader.ReadLineAsync()) != null) Handle(line, c);
        }
        catch { /* dropped */ }
        finally { lock (_conns) _conns.Remove(c); }
    }

    private void Handle(string line, TcpClient c)
    {
        var parts = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
        if (parts.Length < 2) return;
        if (parts[0] == "CLAIM" && parts[1] != _myName)
        {
            var ip = (c.Client.RemoteEndPoint as IPEndPoint)?.Address.ToString() ?? _peerIp ?? "";
            Log?.Invoke($"peer \"{parts[1]}\" claimed control");
            PeerClaimedControl?.Invoke(parts[1], ip);
        }
    }

    /// <summary>Tell peers that this machine is now the active controller.</summary>
    public void BroadcastClaim()
    {
        List<TcpClient> snapshot;
        lock (_conns) snapshot = _conns.ToList();
        foreach (var c in snapshot) SendLine(c, $"CLAIM {_myName}");
    }

    private void SendLine(TcpClient c, string s)
    {
        try { var b = Encoding.UTF8.GetBytes(s + "\n"); c.GetStream().Write(b, 0, b.Length); }
        catch { /* dropped */ }
    }

    public void Dispose()
    {
        _cts?.Cancel();
        try { _listener?.Stop(); } catch { }
        lock (_conns) { foreach (var c in _conns) c.Dispose(); _conns.Clear(); }
    }
}

File: cozy-desktop/macos/CozyMac/PeerCoordinator.swift
import Foundation
import Network

/// Control channel between Cozy desktops (PC ↔ Mac) for multi-controller handoff.
/// Mirrors the Windows PeerCoordinator: listens on a control port, dials the peer,
/// and exchanges HELLO/CLAIM lines. When a peer claims control, we yield.
final class PeerCoordinator {
    static let controlPort: UInt16 = 24811

    var onPeerClaim: ((_ peerName: String, _ peerIp: String) -> Void)?
    var onLog: ((String) -> Void)?
    private(set) var peerName: String = ""

    private var listener: NWListener?
    private var connections: [NWConnection] = []
    private var myName = ""
    private var peerHost: String?

    func start(myName: String, peerHost: String?) {
        self.myName = myName
        self.peerHost = (peerHost?.isEmpty ?? true) ? nil : peerHost
        startListener()
        if let h = self.peerHost { add(NWConnection(host: NWEndpoint.Host(h),
                                                     port: NWEndpoint.Port(rawValue: Self.controlPort)!,
                                                     using: .tcp)) }
    }

    private func startListener() {
        do {
            let l = try NWListener(using: .tcp, on: NWEndpoint.Port(rawValue: Self.controlPort)!)
            l.newConnectionHandler = { [weak self] conn in self?.add(conn) }
            l.start(queue: .global())
            listener = l
        } catch { onLog?("coordinator listen failed: \(error)") }
    }

    private func add(_ conn: NWConnection) {
        connections.append(conn)
        conn.start(queue: .global())
        send(conn, "HELLO \(myName)")
        receive(conn)
    }

    private func receive(_ conn: NWConnection) {
        conn.receive(minimumIncompleteLength: 1, maximumLength: 4096) { [weak self] data, _, isComplete, _ in
            if let data = data, let s = String(data: data, encoding: .utf8) {
                for line in s.split(separator: "\n") { self?.handle(String(line), conn) }
            }
            if isComplete { conn.cancel() } else { self?.receive(conn) }
        }
    }

    private func handle(_ line: String, _ conn: NWConnection) {
        let parts = line.split(separator: " ").map(String.init)
        guard parts.count >= 2 else { return }
        switch parts[0] {
        case "HELLO":
            peerName = parts[1]
        case "CLAIM" where parts[1] != myName:
            // For a 2-machine setup each side configures the other's IP, so peerHost
            // is the correct address to receive from.
            let ip = peerHost ?? ""
            onLog?("peer \"\(parts[1])\" claimed control")
            onPeerClaim?(parts[1], ip)
        default:
            break
        }
    }

    func broadcastClaim() { for c in connections { send(c, "CLAIM \(myName)") } }

    private func send(_ conn: NWConnection, _ s: String) {
        conn.send(content: (s + "\n").data(using: .utf8), completion: .contentProcessed { _ in })
    }

    func stop() {
        listener?.cancel()
        listener = nil
        connections.forEach { $0.cancel() }
        connections.removeAll()
    }
}
